File: gapic/src/service/com/google/gapid/service/atom/Observations.java
/*
 * Copyright (C) 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.gapid.service.atom;

import com.google.gapid.rpclib.binary.BinaryClass;
import com.google.gapid.rpclib.binary.BinaryObject;
import com.google.gapid.rpclib.binary.Decoder;
import com.google.gapid.rpclib.binary.Encoder;
import com.google.gapid.rpclib.binary.Namespace;
import com.google.gapid.rpclib.schema.Entity;
import com.google.gapid.rpclib.schema.Field;
import com.google.gapid.rpclib.schema.Slice;
import com.google.gapid.rpclib.schema.Struct;

import java.io.IOException;
import java.util.Arrays;

public final class Observations implements BinaryObject {

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Observations that = (Observations)o;
    if (!Arrays.equals(myReads, that.myReads)) return false;
    if (!Arrays.equals(myWrites, that.myWrites)) return false;
    return true;
  }

  @Override
  public int hashCode() {
    int result = Arrays.hashCode(myReads);
    result = 31 * result + Arrays.hashCode(myWrites);
    return result;
  }

  @Override
  public String toString() {
    return "Observations{reads=" + Arrays.toString(myReads) + ", writes=" + Arrays.toString(myWrites) + '}';
  }

  //<<<Start:Java.ClassBody:1>>>
  private Observation[] myReads;
  private Observation[] myWrites;

  // Constructs a default-initialized {@link Observations}.
  public Observations() {}


  public Observation[] getReads() {
    return myReads;
  }

  public Observations setReads(Observation[] v) {
    myReads = v;
    return this;
  }

  public Observation[] getWrites() {
    return myWrites;
  }

  public Observations setWrites(Observation[] v) {
    myWrites = v;
    return this;
  }

  @Override
  public BinaryClass klass() { return Klass.INSTANCE; }


  private static final Entity ENTITY = new Entity("atom", "Observations", "", "");

  static {
    ENTITY.setFields(new Field[]{
      new Field("Reads", new Slice("", new Struct(Observation.Klass.INSTANCE.entity()))),
      new Field("Writes", new Slice("", new Struct(Observation.Klass.INSTANCE.entity()))),
    });
    Namespace.register(Klass.INSTANCE);
  }
  public static void register() {}
  //<<<End:Java.ClassBody:1>>>
  public enum Klass implements BinaryClass {
    //<<<Start:Java.KlassBody:2>>>
    INSTANCE;

    private static final Observation[] Observation_EMPTY = {};

    @Override
    public Entity entity() { return ENTITY; }

    @Override
    public BinaryObject create() { return new Observations(); }

    @Override
    public void encode(Encoder e, BinaryObject obj) throws IOException {
      Observations o = (Observations)obj;
      e.uint32(o.myReads.length);
      for (int i = 0; i < o.myReads.length; i++) {
        e.value(o.myReads[i]);
      }
      e.uint32(o.myWrites.length);
      for (int i = 0; i < o.myWrites.length; i++) {
        e.value(o.myWrites[i]);
      }
    }

    @Override
    public void decode(Decoder d, BinaryObject obj) throws IOException {
      Observations o = (Observations)obj;
      {
        int len = d.uint32();
        o.myReads = len != 0 ? new Observation[len] : Observation_EMPTY;
      }
      for (int i = 0; i <o.myReads.length; i++) {
        o.myReads[i] = new Observation();
        d.value(o.myReads[i]);
      }
      {
        int len = d.uint32();
        o.myWrites = len != 0 ? new Observation[len] : Observation_EMPTY;
      }
      for (int i = 0; i <o.myWrites.length; i++) {
        o.myWrites[i] = new Observation();
        d.value(o.myWrites[i]);
      }
    }
    //<<<End:Java.KlassBody:2>>>
  }
}

File: gapis/atom/observations.go
// Copyright (C) 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package atom

import (
	"bytes"
	"context"
	"fmt"

	"github.com/google/gapid/core/data/id"
	"github.com/google/gapid/core/data/protoconv"
	"github.com/google/gapid/gapis/database"
	"github.com/google/gapid/gapis/memory"
	"github.com/google/gapid/gapis/memory/memory_pb"
)

// Observations is a collection of reads and write observations performed by an
// atom.
type Observations struct {
	Reads  []Observation
	Writes []Observation
}

func (o *Observations) String() string {
	return fmt.Sprintf("Reads: %v, Writes: %v", o.Reads, o.Writes)
}

// AddRead appends the read to the list of observations.
func (o *Observations) AddRead(rng memory.Range, id id.ID) {
	o.Reads = append(o.Reads, Observation{Range: rng, ID: id})
}

// AddWrite appends the write to the list of observations.
func (o *Observations) AddWrite(rng memory.Range, id id.ID) {
	o.Writes = append(o.Writes, Observation{Range: rng, ID: id})
}

// ApplyReads applies all the observed reads to memory pool p.
// This is a no-op when called when o is nil.
func (o *Observations) ApplyReads(p *memory.Pool) {
	if o != nil {
		for _, r := range o.Reads {
			p.Write(r.Range.Base, memory.Resource(r.ID, r.Range.Size))
		}
	}
}

// ApplyReads applies all the observed writes to the memory pool p.
// This is a no-op when called when o is nil.
func (o *Observations) ApplyWrites(p *memory.Pool) {
	if o != nil {
		for _, w := range o.Writes {
			p.Write(w.Range.Base, memory.Resource(w.ID, w.Range.Size))
		}
	}
}

// DataString returns a string describing all reads/writes and their raw data.
func (o *Observations) DataString(ctx context.Context) string {
	var buf bytes.Buffer
	for _, read := range o.Reads {
		buf.WriteString(fmt.Sprintf("[read] %v\n", read))
		if data, err := database.Resolve(ctx, read.ID); err == nil {
			buf.WriteString(fmt.Sprintf("[data] %v\n", data))
		} else {
			buf.WriteString(fmt.Sprintf("[data] failed: %v\n", err))
		}
	}
	for _, write := range o.Writes {
		buf.WriteString(fmt.Sprintf("[write] %v\n", write))
		if data, err := database.Resolve(ctx, write.ID); err == nil {
			buf.WriteString(fmt.Sprintf("[data] %v\n", data))
		} else {
			buf.WriteString(fmt.Sprintf("[data] failed: %v\n", err))
		}
	}
	return buf.String()
}

// Observation represents a single read or write observation made by an atom.
type Observation struct {
	Range memory.Range // Memory range that was observed.
	ID    id.ID        // The resource identifier of the observed data.
}

func (o Observation) String() string {
	return fmt.Sprintf("{Range: %v, ID: %v}", o.Range, o.ID)
}

func init() {
	protoconv.Register(
		func(ctx context.Context, a Observation) (*memory_pb.Observation, error) {
			return &memory_pb.Observation{
				Base: a.Range.Base,
				Size: a.Range.Size,
				Id:   a.ID.String(),
			}, nil
		},
		func(ctx context.Context, a *memory_pb.Observation) (Observation, error) {
			o := Observation{}
			o.Range.Base = a.Base
			o.Range.Size = a.Size
			o.ID.Parse(a.Id)
			return o, nil
		},
	)
}
